prevents garbage collection from removing pinned images

wgahnagl opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
prevents garbage collection from removing pinned images
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

https://github.com/kubernetes/kubernetes/pull/101808 
implements this
#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
allows users to prevent garbage collection on pinned images 

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:


- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP] https://github.com/kubernetes/enhancements/pull/2694
```

ehashman commented:
Is there a KEP or issue tracking this?

haircommander commented:
this was combined with CRI beta kep: https://github.com/kubernetes/enhancements/pull/2694

haircommander commented:
/retest

wgahnagl commented:
/priority important-soon

endocrimes commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -338,13 +338,29 @@ func (im *realImageGCManager) freeSpace(bytesToFree int64, freeTime time.Time) (
+	if err != nil {
+		return 0, err
+	}
+	imagesMap := make(map[string]container.Image)
```
nit: Preallocate the capacity of the map? - Negligible gain in most cases, but hosts can end up with a lot of images so it's probably worth it

wgahnagl replied:
how big should I make it ?

wgahnagl replied:
fixed!

endocrimes commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -338,13 +338,29 @@ func (im *realImageGCManager) freeSpace(bytesToFree int64, freeTime time.Time) (
 		}
+		// Check if image is pinned, prevent garbage collection
+		if imagesMap[image].Pinned {
+			klog.V(5).InfoS("Image is pinned", "imageID", image)
```
```suggestion
			klog.V(5).InfoS("Image is pinned, skipping garbage collection", "imageID", image)
```

endocrimes commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1130,6 +1130,10 @@ message Image {
     // ImageSpec for image which includes annotations
     ImageSpec spec = 7;
+    // Recommendation on whether this image should be exempt from garbage collection.
+    // It must only be treated as a recommendation--the client can still request the image be deleted,
```
```suggestion
    // It must only be treated as a recommendation -- the client can still request that the image be deleted,
```

wgahnagl replied:
fixed!

endocrimes commented on pkg/kubelet/images/image_gc_manager_test.go:
```diff
@@ -193,6 +193,25 @@ func TestDeleteUnusedImagesExemptSandboxImage(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestDeletePinnedImage(t *testing.T) {
```
I'd recommend adding cases for:
- List with no pinned images
- List with only pinned images
- List with mixed images

wgahnagl replied:
added!

wgahnagl commented:
/retest

endocrimes commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -338,13 +338,29 @@ func (im *realImageGCManager) freeSpace(bytesToFree int64, freeTime time.Time) (
+		return 0, err
+	}
+	imagesMap := make(map[string]container.Image, len(allImages))
+	for i := 0; i < len(allImages); i += 1 {
```
is there a reason you aren't iterating over the list directly here?

wgahnagl replied:
I got a different review a while ago that said to preallocate the map, because it helped performance on large lists of images

endocrimes replied:
Fortunately that doesn't stop you from iterating - Which is nice for not setting off "this seems weird, lets look closer" alarms in people's heads.

```
for idx, img := range allImages
```

ehashman reviewed (changes requested):
Please fix the verify failures.

/triage accepted

haircommander commented:
/lgtm

thanks for sticking with this @wgahnagl !

wgahnagl commented:
@feiskyer @andrewsykim ptal

wgahnagl commented:
@endocrimes  @mrunalp ptal !

brandond commented:
We're excited to see this land.

pacoxu commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -338,13 +338,29 @@ func (im *realImageGCManager) freeSpace(bytesToFree int64, freeTime time.Time) (
+	for _, img := range allImages {
+		imagesMap[img.ID] = img
+	}
+
```
After reading the code, I think another way is adding pinned for imageRecords and set it in `func detectImages()`. 

There is an iterator in detectImages. We can iterate once. Like image.size. 
https://github.com/kubernetes/kubernetes/blob/10976cbb43022c01a7ece07079cb9cedf04aeb90/pkg/kubelet/images/image_gc_manager.go#L240-L260

wgahnagl committed:
cr-api: add pinned field to Image

For the CRI implementation to be able to indicate to the Kubelet that an image should not be garbage collected.

Signed-off-by: Peter Hunt <pehunt@redhat.com>

wgahnagl committed:
prevents garbage collection from removing pinned images

wgahnagl committed:
adds pinned field to imageRecords

wgahnagl commented:
@pacoxu sorry this took so long! Ptal!

endocrimes commented on pkg/kubelet/images/image_gc_manager_test.go:
```diff
@@ -206,6 +206,105 @@ func TestDeleteUnusedImagesExemptSandboxImage(t *testing.T) {
 
+func TestDeletePinnedImage(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
```
no need to call `Finish` when passing a testing.T into NewController as of go1.14 (https://pkg.go.dev/github.com/golang/mock/gomock#Controller.Finish)

```suggestion

```

endocrimes reviewed (commented):
logic lgtm, minor test cleanup comments

pacoxu reviewed (commented):
/lgtm
Seems good now.
